torchvision/csrc/ops/cuda/roi_pool_kernel.cu

```diff
@@ -113,10 +114,13 @@ __global__ void roi_pool_backward_kernel_impl(
     int argmax = argmax_data_offset[ph * pooled_width + pw];
 
     if (argmax != -1) {
-      gpuAtomicAdd(
-          grad_input_offset + argmax,
-          static_cast<T>(
-              grad_output[output_offset + ph * h_stride + pw * w_stride]));
+      // memory_span == grad_input.numel()
+      // grad_input + index, index < memory_span
+      // grad_input_offset + j == grad_input + offset + j, j < numel
+      // -> offset + j < memory_span => j < memory_span - offset
+      const int numel = memory_span - ((roi_batch_ind * channels + c) * height * width);
```

Comment:
It would be better to not be recomputing the numel. Is it inconvenient to do it properly here?

Reply:
I'm not very sure about that. If i understand correctly `fastAtomicAdd` API, numel argument (3rd arg) is the boundary for provided index (2nd arg). As provious code was using `grad_input_offset` which is `grad_input` plus an offset (`((roi_batch_ind * channels + c) * height * width)`), we have to recompute the boundary for given index.

Maybe, we can use `grad_input` pointer and put the offset into index argument, so that numel will be `memory_span`:
```
auto offset = ((roi_batch_ind * channels + c) * height * width);
at::native::fastAtomicAdd(grad_input, offset + argmax, memory_span, ...);
```

Reply:
OK, so I took a closer look at fast atomic add to understand why it needs the numel, and the answer is that the fast atomic add trick is solely for half precision additions, and works by turning a half-size atomic (which is slow) into a double size half2 atomic (which is fast), where the half (ha!) of half2 which is not updated is zeroed. However, if the atomic update is at the very bounds of the tensor in question, we're not allowed to do this, since it may not be valid to do a memory access out of bounds.

This means that we don't need the "true" numel of the tensor in question, we only need to approximately what is out of bounds or not. Similarly, you want the base pointer to be the very beginning of the tensor.

Does that make sense?

Reply:
Thanks for the explanation, makes sense! I updated the PR, if you can review it once again, thanks